### ISSUE TITLE
Custom properties on schema property configuration object

### ISSUE THREAD

**Issue by govindrai** [Automattic/mongoose#7303]
Is there a way to add custom properties to schema property's configuration object? i.e. I would like to denote properties as "cloud only" and want to pull them up later when doing validation to ensure they aren't changed by the client (but I still want to be able to update them). Right now I am having to create a list of cloud only properties in an array, but this poses a problem in the event I forget to update that array. I tried to add `cloudOnly: true` just to try it but run into `TypeError: Undefined type `undefined` at `tokens.bru.cloudOnly``. This was expected. Do you have any solutions for this?

Here's an example: 

```
const userSchema = new mongoose.Schema({
  userId: {
    type: mongoose.ObjectId,
    required: true,
    cloudOnly: true,
  }
});
```

**Comment by govindrai** [comment]
@mhombach Glad you understand my issue! I was hoping Mongoose would just ignore that property but allow it to exist on the object, but it errors out. Yea, I'll wait for @lineus or @vkarpov15 to hear their suggestion..

**Comment by vkarpov15** [comment]
@govindrai I can't repro your issue, the below script works fine for me. @mhombach 's explanation is unfortunately incorrect.

```javascript
const assert = require('assert');
const mongoose = require('mongoose');
mongoose.set('debug', true);

const GITHUB_ISSUE = `gh7303`;
const connectionString = `mongodb://localhost:27017/${ GITHUB_ISSUE }`;
const { Schema } = mongoose;

run().then(() => console.log('done')).catch(error => console.error(error.stack));

async function run() {
  await mongoose.connect(connectionString);
  await mongoose.connection.dropDatabase();

  const userSchema = new mongoose.Schema({
    userId: {
      type: mongoose.ObjectId,
      required: true,
      cloudOnly: true,
    }
  });
}
```

In general, there's no reason why you can't add custom properties on your schema types. There are plenty of plugins that rely on custom schematype options, like [mongoose-autopopulate](http://plugins.mongoosejs.io/plugins/autopopulate). Try using [`User.schema.eachPath()`](https://mongoosejs.com/docs/api.html#schema_Schema-eachPath) to iterate through every path in the user model's schema. Here's an [example](https://github.com/mongodb-js/mongoose-autopopulate/blob/4ccae73c11c5b28dbc112cd22ae8dddb08343113/index.js#L134)

**Comment by vkarpov15** [comment]
No worries, we don't really have docs in place to explain that it is perfectly fine to have arbitrary options on a schematype. Something to add.

What typescript IDE are you using? It wouldn't be the first time that the Typescript bindings are out of date or not 100% correct.

Random thought: would you use officially supported mongoose TS bindings?

**Comment by govindrai** [comment]
@vkarpov15 You're absolutely correct. It's definitely works. For whatever reason, I couldn't understand the error message. 

Instead of add an option to a path's configuration object, I accidentally added `cloudOnly` as an actual path with a value of `true`. It was a pretty nested object so looking at it didn't register... 😓 Hence the error `TypeError: Undefined type 'undefined' at 'tokens.bru.cloudOnly'`.  Maybe an `invalid schema configuration defined for tokens.bru.cloudOnly` would be a better message?

i.e.
```
const userSchema = new mongoose.Schema({
  userId: {
    type: mongoose.ObjectId,
    required: true
  },
  cloudOnly: true,
});
```


Also never knew about `Schema.prototype.eachPath()`! Will check it out. Thanks for opening up that issue for the clarification.

**Comment by vkarpov15** [comment]
Good idea, we will improve this error message
